### PR TITLE
[Feat] Add download status into runtime downloader

### PR DESCRIPTION
**Pull request by brosoul**
## Pull Request Description
In the PR, the status of model download `DownloadFile().status` is provided, and a filelock is used to lock the downloading files to avoid unknown errors caused by multiple processes downloading the same model file.

image

## Related Issues
Resolves: #454
Part of #521 (Visibility of all models)

**Important: Before submitting, please complete the description above and review the checklist below.**

---

<details>
<summary><strong>Contribution Guidelines (Expand for Details)</strong></summary>

<p>We appreciate your contribution to aibrix! To ensure a smooth review process and maintain high code quality, please adhere to the following guidelines:</p>

<h3>Pull Request Title Format</h3>
<p>Your PR title should start with one of these prefixes to indicate the nature of the change:</p>
<ul>
    <li><code>[Bug]</code>: Corrections to existing functionality</li>
    <li><code>[CI]</code>: Changes to build process or CI pipeline</li>
    <li><code>[Docs]</code>: Updates or additions to documentation</li>
    <li><code>[API]</code>: Modifications to aibrix's API or interface</li>
    <li><code>[CLI]</code>: Changes or additions to the Command Line Interface</li>
    <li><code>[Misc]</code>: For changes not covered above (use sparingly)</li>
</ul>
<p><em>Note: For changes spanning multiple categories, use multiple prefixes in order of importance.</em></p>

<h3>Submission Checklist</h3>
<ul>
    <li>[ ] PR title includes appropriate prefix(es)</li>
    <li>[ ] Changes are clearly explained in the PR description</li>
    <li>[ ] New and existing tests pass successfully</li>
    <li>[ ] Code adheres to project style and best practices</li>
    <li>[ ] Documentation updated to reflect changes (if applicable)</li>
    <li>[ ] Thorough testing completed, no regressions introduced</li>
</ul>

<p>By submitting this PR, you confirm that you've read these guidelines and your changes align with the project's contribution standards.</p>

</details>

### CODE REVIEW

File: python/aibrix/aibrix/config.py [hunk 1]
```diff
@@ -14,7 +14,15 @@
...
-DOWNLOAD_CACHE_DIR = ".cache"
+"""
+DOWNLOAD CACHE DIR would be like:
+.
+└── .cache
+    └── huggingface | s3 | tos
+        ├── .gitignore
+        └── download
+"""
+DOWNLOAD_CACHE_DIR = ".cache/%s/download"
```
Comment (Jeffwan) [inline python/aibrix/aibrix/config.py:25]: do we need to distinguish the source of the artifacts? that means S3 downloaded models can not be searched by TOS loader? is that just for safe or there's some minor difference (like file structure) between sources?

File: python/aibrix/aibrix/downloader/entity.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+class RemoteSource(Enum):
+    S3 = "s3"
+    TOS = "tos"
+    HUGGINGFACE = "huggingface"
+
+
+class DownloadStatus(Enum):
+    DOWNLOADING = "downloading"
+    DOWNLOADED = "downloaded"
+    NO_OPERATION = "no_operation"  # Interrupted from downloading
```
Comment (Jeffwan) [inline python/aibrix/aibrix/downloader/entity.py:46]: If 1 process is downloading the model, the other process should wait there, what's the file status? no_operation? beside above case, any other case will be `no_operation`?

File: python/aibrix/aibrix/downloader/entity.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+logger = init_logger(__name__)
+
+
+class RemoteSource(Enum):
+    S3 = "s3"
+    TOS = "tos"
+    HUGGINGFACE = "huggingface"
+
+
+class DownloadStatus(Enum):
```
Comment (Jeffwan) [inline python/aibrix/aibrix/downloader/entity.py]: After reading testing code, I am a little bit confused here. `DownloadStatus` seems designed for both `File` and `Model`. are their state machine exact same? If not, I suggest to extend to `FileDownloadStatus` or `ModelDownloadStatus` etc

File: python/aibrix/aibrix/downloader/entity.py (hunk 2 above)
Comment (brosoul) [reply python/aibrix/aibrix/downloader/entity.py:46]: > If 1 process is downloading the model, the other process should wait there, what's the file status?

The file status is `downloading`.
Only the downloading processing.

Only when the unique download process of the model exits due to an exception, will the model file be in the 'no_operation' state

File: python/aibrix/aibrix/downloader/entity.py (hunk 3 above)
Comment (brosoul) [reply python/aibrix/aibrix/downloader/entity.py]: > extend to FileDownloadStatus or ModelDownloadStatus

This seems quite reasonable

File: python/aibrix/aibrix/config.py (hunk 1 above)
Comment (brosoul) [reply python/aibrix/aibrix/config.py:25]: TOS and S3 have the same structure, but we are concerned that we may need to identify which download source this model comes from in the future. So the current directory structure has been retained.

File: python/aibrix/aibrix/downloader/entity.py (hunk 3 above)
Comment (brosoul) [reply python/aibrix/aibrix/downloader/entity.py]: Please take a look

Comment (Jeffwan) [reply python/aibrix/aibrix/downloader/entity.py]: great

File: python/aibrix/aibrix/downloader/entity.py (hunk 2 above)
Comment (Jeffwan) [reply python/aibrix/aibrix/downloader/entity.py:46]: Sounds good.

**Pull request merged**